### PR TITLE
heap store.get directly calls map.get #787

### PR DESCRIPTION
**Pull request by AbfrmBlr**

### CODE REVIEW

File: core/src/main/java/org/ehcache/core/spi/cache/AbstractValueHolder.java [hunk 1]
```diff
@@ -109,7 +124,15 @@ public long lastAccessTime(TimeUnit unit) {
   }
 
   public void setLastAccessTime(long lastAccessTime, TimeUnit unit) {
-    this.lastAccessTime = unit.convert(lastAccessTime, nativeTimeUnit());
+    while (true) {
```
Comment (lorban) [inline core/src/main/java/org/ehcache/core/spi/cache/AbstractValueHolder.java]: Why a while loop here and a do...while loop above?

**Comment by AbfrmBlr** [comment]
I think for `lastaccesstime` loop is not required. removing that.

**Comment by AbfrmBlr** [comment]
That was stupid. Loop is required for both.

**Comment by AbfrmBlr** [comment]
`setLastAccessTime` had a bug - resolved now

File: core/src/main/java/org/ehcache/core/spi/cache/AbstractValueHolder.java [hunk 2]
```diff
@@ -109,7 +124,16 @@ public long lastAccessTime(TimeUnit unit) {
...
 
   public void setLastAccessTime(long lastAccessTime, TimeUnit unit) {
-    this.lastAccessTime = unit.convert(lastAccessTime, nativeTimeUnit());
+    lastAccessTime = unit.convert(lastAccessTime, nativeTimeUnit());
+    do {
+      long read = this.lastAccessTime;
+      if (read >= lastAccessTime) {
+        break;
+      }
+      if (ACCESSTIME_UPDATER.compareAndSet(this, read, unit.convert(lastAccessTime, nativeTimeUnit()))) {
```
Comment (lorban) [inline core/src/main/java/org/ehcache/core/spi/cache/AbstractValueHolder.java]: `lastAccessTime` has already been converted, don't do it again.

File: core/src/main/java/org/ehcache/core/spi/cache/AbstractValueHolder.java [hunk 3]
```diff
@@ -109,7 +124,16 @@ public long lastAccessTime(TimeUnit unit) {
   }
 
   public void setLastAccessTime(long lastAccessTime, TimeUnit unit) {
-    this.lastAccessTime = unit.convert(lastAccessTime, nativeTimeUnit());
+    lastAccessTime = unit.convert(lastAccessTime, nativeTimeUnit());
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/spi/cache/AbstractValueHolder.java]: Removing the `this.` does not magically create a local var. You are still setting it blindly.

Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/spi/cache/AbstractValueHolder.java]: hehe scratch that, you are assigning to the param. Well, could we have a local var with a name instead? This is confusing with the name shadowing in place already.

File: core/src/main/java/org/ehcache/core/spi/cache/AbstractValueHolder.java [hunk 4]
```diff
@@ -54,14 +57,26 @@ public long creationTime(TimeUnit unit) {
...
+  private void updateExpirationTime(long update) {
+    do {
+      long current = this.expirationTime;
+      if (current >= update) {
+        break;
+      }
+      if (EXPIRATIONTIME_UPDATER.compareAndSet(this, current, update)) {
+        break;
+      }
+    } while (true);
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/spi/cache/AbstractValueHolder.java]: Given the condition, I am in favour of a `while(true)` loop that has the benefit of clearly indicating you are doing something special.

File: core/src/main/java/org/ehcache/core/spi/cache/AbstractValueHolder.java [hunk 5]
```diff
@@ -109,7 +124,16 @@ public long lastAccessTime(TimeUnit unit) {
...
+    lastAccessTime = unit.convert(lastAccessTime, nativeTimeUnit());
+    do {
+      long read = this.lastAccessTime;
+      if (read >= lastAccessTime) {
+        break;
+      }
+      if (ACCESSTIME_UPDATER.compareAndSet(this, read, unit.convert(lastAccessTime, nativeTimeUnit()))) {
+        break;
+      }
+    } while (true);
```
Comment (ljacomet) [inline core/src/main/java/org/ehcache/core/spi/cache/AbstractValueHolder.java]: Same as for the other `do ... while`

File: core/src/main/java/org/ehcache/core/spi/cache/AbstractValueHolder.java [hunk 6]
```diff
@@ -109,7 +124,16 @@ public long lastAccessTime(TimeUnit unit) {
   }
 
   public void setLastAccessTime(long lastAccessTime, TimeUnit unit) {
-    this.lastAccessTime = unit.convert(lastAccessTime, nativeTimeUnit());
+    lastAccessTime = unit.convert(lastAccessTime, nativeTimeUnit());
+    while (true) {
+      long read = this.lastAccessTime;
```
Comment (lorban) [inline core/src/main/java/org/ehcache/core/spi/cache/AbstractValueHolder.java]: Renaming the `read` variable `current` and the local `lastAccessTime` one `update` would make this method trivial to mentally match it with the `updateExpirationTime()` one. That would help future maintenance.

File: impl/src/main/java/org/ehcache/impl/internal/store/heap/OnHeapStore.java [hunk 7]
```diff
@@ -263,36 +263,29 @@ public OnHeapStore(final Configuration<K, V> config, final TimeSource timeSource
...
-        if (expiredValue.get() != null) {
-          decrementCurrentUsageInBytesIfRequired(expiredValue.get().size());
-        }
-      } else {
+        storeEventDispatcher.releaseEventSink(eventSink);
+        return null;
+      }
+
+      if (updateAccess) {
+        mapping = setAccessTimeAndExpiryThenReturnMapping(key, mapping, timeSource.getTimeMillis(), eventSink);
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/internal/store/heap/OnHeapStore.java]: This is the only place we need the event sink, which means it should be created and released around this call.

File: impl/src/main/java/org/ehcache/impl/internal/store/heap/OnHeapStore.java [hunk 8]
```diff
@@ -263,36 +263,29 @@ public OnHeapStore(final Configuration<K, V> config, final TimeSource timeSource
...
+      }
+      if (mapping != null) {
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/internal/store/heap/OnHeapStore.java]: You are missing the else branch here

**Pull request merged**